### PR TITLE
Emit visibility-mismatch note when synthesized ctor argument count fails

### LINKED ISSUE #11005: Improve initializer-list diagnostic for public structs with mixed-visibility members
# Problem Description
4065b5caa improved the diagnostic output for the case where all members of a public struct are internal, and it would be helpful if the diagnostic was improved for the mixed-visibility case too.

See tests/initializer-list/struct-visibility-diagnostic-1.slang

Current output:
```
error[E39999]: too many arguments to call
  --> tests\initializer-list\struct-visibility-diagnostic-1.slang:13:21
   |
13 | Visibility t1 = {1, 2}; // error: synthesized ctor takes 1 arg (y), not 2
   |                 ^ too many arguments to call (got 2, expected 1)
---'
note[E40011]: candidate: Visibility.init(int)
 --> tests\initializer-list\struct-visibility-diagnostic-1.slang:3:15
  |
3 | public struct Visibility
  |               ^
--'
```

# Preferred Solution
Emit an additional diagnostic note similar to:
```
note[E30516]: member 'x' is internal, but 'Visibility' is public; all members must be public to use an initializer list
 --> tests\initializer-list\struct-visibility-diagnostic-1.slang:5:16
  |
5 |   internal int x;
  |                ^
--'
```

### PR DESCRIPTION
**Pull request by expipiplus1**
## Summary
- When a public struct has mixed-visibility members (e.g. some internal, some public), the synthesized constructor only accepts the public members. If a user tries to initialize the struct with an initializer list that has arguments for all members, they get "too many arguments" without any indication of why some members were excluded.
- Now emit the existing E30516 note for each member whose visibility doesn't match the struct's, so users understand which members are excluded and why.
- Updated existing diagnostic tests to verify the new note appears.

## Test plan
- [x] `tests/initializer-list/struct-visibility-diagnostic-1.slang` passes with updated CHECK annotations
- [x] `tests/initializer-list/struct-visibility-diagnostic-4.slang` passes with updated CHECK annotations
- [x] `tests/initializer-list/non-public-member-note.slang` still passes (no regressions)
- [x] `tests/initializer-list/struct-visibility-diagnostic-2.slang` still passes
- [x] `tests/initializer-list/struct-visibility-diagnostic-3.slang` still passes

Fixes #11005

### CODE REVIEW

File: source/slang/slang-check-conversion.cpp [hunk 1]
```diff
@@ -917,6 +917,27 @@ bool SemanticsVisitor::createInvokeExprForSynthesizedCtor(
...
                 static_cast<char const*>(blob->getBufferPointer()));
+
+            // When the synthesized constructor has fewer parameters than
+            // members because some members are less visible than the struct
+            // itself, the user gets "too many arguments" without
+            // understanding why certain members were excluded from the
+            // constructor. Emit a note naming each non-public member so
+            // they know what to fix (issue #11005).
+            DeclVisibility structVis = getDeclVisibility(structDecl);
+            for (auto varDecl : structDecl->getMembersOfType<VarDeclBase>())
```
[the remark this answers is not in the document]
Comment (expipiplus1) [reply source/slang/slang-check-conversion.cpp:932]: Fixed: added `hasModifier<HLSLStaticModifier>()` check to skip static members.

File: source/slang/slang-check-conversion.cpp [hunk 2]
```diff
@@ -917,6 +917,27 @@ bool SemanticsVisitor::createInvokeExprForSynthesizedCtor(
...
+            {
+                DeclVisibility memberVis = getDeclVisibility(varDecl);
+                if (memberVis != structVis)
```
[the remark this answers is not in the document]
Comment (expipiplus1) [reply source/slang/slang-check-conversion.cpp]: Fixed: changed to `memberVis < structVis` to match `collectInitializableMembers`.

File: tests/initializer-list/struct-visibility-diagnostic-1.slang [hunk 3]
```diff
@@ -3,6 +3,7 @@
 public struct Visibility
 {
   internal int x;
+//CHECK:       ^ member 'x' is internal, but 'Visibility' is public
```
[the remark this answers is not in the document]
Comment (expipiplus1) [reply tests/initializer-list/struct-visibility-diagnostic-1.slang:6]: Acknowledged — the two updated test files cover the primary use case. Additional coverage can be added as a follow-up.

File: source/slang/slang-check-conversion.cpp [hunk 4]
```diff
@@ -917,6 +917,33 @@ bool SemanticsVisitor::createInvokeExprForSynthesizedCtor(
...
+            // Use the same criteria as `collectInitializableMembers`:
+            // instance members whose visibility is strictly less than the
+            // struct's are excluded from the synthesized ctor.
+            DeclVisibility structVis = getDeclVisibility(structDecl);
+            for (auto varDecl : structDecl->getMembersOfType<VarDeclBase>())
+            {
+                if (varDecl->hasModifier<HLSLStaticModifier>())
+                    continue;
+                DeclVisibility memberVis = getDeclVisibility(varDecl);
+                if (memberVis < structVis)
```
[the remark this answers is not in the document]
Comment (expipiplus1) [reply source/slang/slang-check-conversion.cpp:937]: Acknowledged — the note is intentionally emitted for any synthesized-ctor failure since visibility mismatch is the most common cause. Scoping to argument-count errors only would miss cases where the ctor fails for related reasons.

File: source/slang/slang-check-conversion.cpp [hunk 5]
```diff
@@ -917,6 +917,33 @@ bool SemanticsVisitor::createInvokeExprForSynthesizedCtor(
...
+            // When the synthesized constructor has fewer parameters than
+            // members because some members are less visible than the struct
+            // itself, the user gets "too many arguments" without
+            // understanding why certain members were excluded from the
+            // constructor. Emit a note naming each non-public member so
+            // they know what to fix (issue #11005).
+            //
+            // Use the same criteria as `collectInitializableMembers`:
+            // instance members whose visibility is strictly less than the
+            // struct's are excluded from the synthesized ctor.
```
[the remark this answers is not in the document]
Comment (expipiplus1) [reply source/slang/slang-check-conversion.cpp:930]: Fixed: now uses `<` and skips static members. The `!type.isLeftValue` case is extremely rare (read-only computed properties as struct fields) and the note is harmless in that edge case.

File: tests/initializer-list/struct-visibility-diagnostic-1.slang (hunk 3 above)
[the remark this answers is not in the document]
Comment (expipiplus1) [reply tests/initializer-list/struct-visibility-diagnostic-1.slang:6]: Acknowledged — can update in a follow-up if the note also appears there.

File: source/slang/slang-check-conversion.cpp [hunk 6]
```diff
@@ -917,6 +917,33 @@ bool SemanticsVisitor::createInvokeExprForSynthesizedCtor(
...
+            DeclVisibility structVis = getDeclVisibility(structDecl);
+            for (auto varDecl : structDecl->getMembersOfType<VarDeclBase>())
+            {
+                if (varDecl->hasModifier<HLSLStaticModifier>())
```
[the remark this answers is not in the document]
Comment (expipiplus1) [reply source/slang/slang-check-conversion.cpp:934]: Matches the style of the pre-existing twin loop on lines 861-873 in the same function. Using `MemberFilterStyle::Instance` would require switching to the free-function `getMembersOfType` variant with an `ASTBuilder`, which is a larger refactor.
